File: problems/13.priority-queue/golang/priority_queue_test.go
package main

import (
	"reflect"
	"testing"
)

// min Heap
func TestPriorityQueue1(t *testing.T) {
	competitor := func(a, b int) bool { return a < b }
	priorityQueue := NewPriorityQueue([]int{}, competitor)

	priorityQueue.
		insert(15).
		insert(35).
		insert(10).
		insert(25).
		insert(20).
		insert(50).
		insert(40)

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{10, 20, 15, 35, 25, 50, 40}) {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", []int{10, 20, 15, 35, 25, 50, 40}, priorityQueue.getHeap())
	}
}

// min heap with initial slice of heap
func TestPriorityQueue2(t *testing.T) {
	competitor := func(a, b int) bool {
		return a < b
	}
	priorityQueue := NewPriorityQueue([]int{25, 40, 15, 20, 10, 50, 35}, competitor)

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{10, 15, 25, 40, 20, 50, 35}) {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", []int{10, 15, 25, 40, 20, 50, 35}, priorityQueue.getHeap())
	}
}

// max Heap
func TestPriorityQueue3(t *testing.T) {
	competitor := func(a, b int) bool { return a > b }
	priorityQueue := NewPriorityQueue([]int{}, competitor)

	priorityQueue.
		insert(15).
		insert(35).
		insert(10).
		insert(25).
		insert(20).
		insert(50).
		insert(40)

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{50, 25, 40, 15, 20, 10, 35}) {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", []int{50, 25, 40, 15, 20, 10, 35}, priorityQueue.getHeap())
	}
}

// max heap with initial slice of heap
func TestPriorityQueue4(t *testing.T) {
	competitor := func(a, b int) bool {
		return a > b
	}

	priorityQueue := NewPriorityQueue([]int{25, 40, 15, 20, 10, 50, 35}, competitor)

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{50, 25, 40, 20, 10, 15, 35}) {
		t.Errorf("Expected (%v) but got"+
			"(%v)", []int{50, 25, 40, 20, 10, 15, 35}, priorityQueue.getHeap())
	}
}

// min heap peek
func TestPriorityQueuePeek1(t *testing.T) {
	priorityQueue := NewPriorityQueue([]int{}, func(a, b int) bool { return a < b })

	priorityQueue.
		insert(15).
		insert(35).
		insert(10).
		insert(25).
		insert(20).
		insert(50).
		insert(40)

	root, _ := priorityQueue.peek()

	if root != 10 {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", 10, root)
	}

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{20, 35, 15, 40, 25, 50}) {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", []int{20, 35, 15, 40, 25, 50}, priorityQueue.getHeap())
	}
}

// max heap peek
func TestPriorityQueuePeek2(t *testing.T) {
	competitor := func(a, b int) bool { return a > b }
	priorityQueue := NewPriorityQueue([]int{}, competitor)

	priorityQueue.
		insert(15).
		insert(35).
		insert(10).
		insert(25).
		insert(20).
		insert(50).
		insert(40)

	root, _ := priorityQueue.peek()

	if root != 50 {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", 50, root)
	}

	if !reflect.DeepEqual(priorityQueue.getHeap(), []int{35, 25, 40, 15, 20, 10}) {
		t.Errorf(" Expected (%v) but got"+
			"(%v)", []int{35, 25, 40, 15, 20, 10}, priorityQueue.getHeap())
	}
}

File: ts/13.priority-queue/priority-queue.test.ts
import { describe, it, assert } from 'vitest';
import { PriorityQueueImpl } from './PriorityQueue';

describe('priority queue', () => {
    it('making a max heap via priority queue', () => {
        const pq = new PriorityQueueImpl([], (a: number, b: number) => a > b);
        pq.insert(15);
        pq.insert(35);
        pq.insert(10);
        pq.insert(25);
        pq.insert(20);
        pq.insert(50);
        pq.insert(40);
        assert.deepEqual(pq.getHeap(), [50, 25, 40, 15, 20, 10, 35]);

        const pq2 = new PriorityQueueImpl(
            [25, 40, 15, 20, 10, 50, 35],
            (a: number, b: number) => a > b
        );
        assert.deepEqual(pq2.getHeap(), [50, 40, 35, 20, 10, 15, 25]);
    });

    it('making a min heap via priority queue', () => {
        const pq = new PriorityQueueImpl([], (a: number, b: number) => a < b);
        const pq2 = new PriorityQueueImpl(
            [25, 40, 15, 20, 10, 50, 35],
            (a: number, b: number) => a < b
        );

        pq.insert(15);
        pq.insert(35);
        pq.insert(10);
        pq.insert(25);
        pq.insert(20);
        pq.insert(50);
        pq.insert(40);

        assert.deepEqual(pq.getHeap(), [10, 20, 15, 35, 25, 50, 40]);
        assert.deepEqual(pq2.getHeap(), [10, 20, 15, 25, 40, 50, 35]);
    });
});
